src/data/real/ereal.lean

```diff
@@ -715,6 +719,37 @@ begin
   { simp only [top_mul_top, to_real_top, mul_zero] }
 end
 
+protected lemma neg_mul (x y : ereal) : -x * y = -(x * y) :=
+begin
+  induction x using ereal.rec; induction y using ereal.rec,
+  { refl },
+  { rcases lt_trichotomy 0 y with hy | rfl | hy,
+    { rw [bot_mul_coe_of_pos hy, neg_bot, top_mul_coe_of_pos hy] },
+    { rw [coe_zero, mul_zero, mul_zero, neg_zero] },
+    { rw [bot_mul_coe_of_neg hy, neg_bot, neg_top, top_mul_coe_of_neg hy] } },
+  { refl },
+  { rcases lt_trichotomy 0 x with hx | rfl | hx,
+    { rw [coe_mul_bot_of_pos hx, neg_bot, ← coe_neg, coe_mul_bot_of_neg (neg_neg_of_pos hx)], },
+    { rw [coe_zero, zero_mul, neg_zero, zero_mul] },
+    { rw [coe_mul_bot_of_neg hx, neg_top, ← coe_neg, coe_mul_bot_of_pos (neg_pos_of_neg hx)], }, },
+  { norm_cast, exact neg_mul _ _, },
+  { rcases lt_trichotomy 0 x with hx | rfl | hx,
+    { rw [coe_mul_top_of_pos hx, neg_top, ← coe_neg, coe_mul_top_of_neg (neg_neg_of_pos hx)], },
+    { rw [coe_zero, zero_mul, neg_zero, zero_mul] },
+    { rw [coe_mul_top_of_neg hx, neg_bot, ← coe_neg, coe_mul_top_of_pos (neg_pos_of_neg hx)], }, },
+  { refl },
+  { rcases lt_trichotomy 0 y with hy | rfl | hy,
+    { rw [top_mul_coe_of_pos hy, neg_top, bot_mul_coe_of_pos hy] },
+    { rw [coe_zero, mul_zero, mul_zero, neg_zero] },
+    { rw [top_mul_coe_of_neg hy, neg_top, neg_bot, bot_mul_coe_of_neg hy] } },
+  { refl }
+end
```

Comment:
@YaelDillies, what do you think of
```suggestion
/-- Induct on two ereals by performing case splits on the sign of one whenever the other is
infinite. -/
@[elab_as_eliminator]
lemma induction₂ {P : ereal → ereal → Prop}
  (t_t : P ⊤ ⊤)
  (t_p : ∀ x : ℝ, 0 < x → P ⊤ x)
  (t_z : P ⊤ 0)
  (t_n : ∀ x : ℝ, x < 0 → P ⊤ x)
  (t_b : P ⊤ ⊥)
  (p_t : ∀ x : ℝ, 0 < x → P x ⊤)
  (p_b : ∀ x : ℝ, 0 < x → P x ⊥)
  (z_t : P 0 ⊤)
  (h : ∀ x y : ℝ, P x y)
  (z_b : P 0 ⊥)
  (n_t : ∀ x : ℝ, x < 0 → P x ⊤)
  (n_b : ∀ x : ℝ, x < 0 → P x ⊥)
  (b_t : P ⊥ ⊤)
  (b_p : ∀ x : ℝ, 0 < x → P ⊥ x)
  (b_z : P ⊥ 0)
  (b_n : ∀ x : ℝ, x < 0 → P ⊥ x)
  (b_b : P ⊥ ⊥) : ∀ x y, P x y
| ⊥ ⊥ := b_b
| ⊥ (y : ℝ) := by { rcases lt_trichotomy 0 y with hy|rfl|hy, exacts [b_p y hy, b_z, b_n y hy] }
| ⊥ ⊤ := b_t
| (x : ℝ) ⊥ := by { rcases lt_trichotomy 0 x with hx|rfl|hx, exacts [p_b x hx, z_b, n_b x hx] }
| (x : ℝ) (y : ℝ) := h _ _
| (x : ℝ) ⊤ := by { rcases lt_trichotomy 0 x with hx|rfl|hx, exacts [p_t x hx, z_t, n_t x hx] }
| ⊤ ⊥ := t_b
| ⊤ (y : ℝ) := by { rcases lt_trichotomy 0 y with hy|rfl|hy, exacts [t_p y hy, t_z, t_n y hy] }
| ⊤ ⊤ := t_t

protected lemma neg_mul' (x y : ereal) : -x * y = -(x * y) :=
begin
  with_cases { apply @induction₂ (λ x y, -x * y = -(x * y)) }; propagate_tags { try { dsimp only} },
  case [t_t, b_t, t_b, b_b] { all_goals { refl } },
  case [t_z, b_z, z_t, z_b] { all_goals { simp only [zero_mul, mul_zero, neg_zero] } },
  case h : x y { norm_cast, exact neg_mul _ _, },
  case p_b : x hx {
    rw [coe_mul_bot_of_pos hx, neg_bot, ← coe_neg, coe_mul_bot_of_neg (neg_neg_of_pos hx)] },
  case n_b : x hx {
    rw [coe_mul_bot_of_neg hx, neg_top, ← coe_neg, coe_mul_bot_of_pos (neg_pos_of_neg hx)] },
  case p_t : x hx {
    rw [coe_mul_top_of_pos hx, neg_top, ← coe_neg, coe_mul_top_of_neg (neg_neg_of_pos hx)] },
  case n_t : x hx {
    rw [coe_mul_top_of_neg hx, neg_bot, ← coe_neg, coe_mul_top_of_pos (neg_pos_of_neg hx)] },
  case t_p : y hy { rw [top_mul_coe_of_pos hy, neg_top, bot_mul_coe_of_pos hy] },
  case t_n : y hy { rw [top_mul_coe_of_neg hy, neg_top, neg_bot, bot_mul_coe_of_neg hy] },
  case b_p : y hy { rw [bot_mul_coe_of_pos hy, neg_bot, top_mul_coe_of_pos hy] },
  case b_n : y hy { rw [bot_mul_coe_of_neg hy, neg_bot, neg_top, top_mul_coe_of_neg hy] },
end
```

Reply:
Oh, interesting idea. I tried something similar, but with a single variable, and as expected I didn't gain much because I had to treat 25 cases. Can you see whether that also golfs `to_real_mul`?

Reply:
I'll try that in a follow up PR I think, no need to do it here.

bors merge

Reply:
#18481